spring-integration-zmq/src/main/java/org/springframework/integration/zmq/inbound/AbstractZmqMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2002-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.zmq.inbound;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.springframework.integration.endpoint.MessageProducerSupport;
+import org.springframework.integration.support.management.IntegrationManagedResource;
+import org.springframework.integration.zmq.support.DefaultZmqMessageConverter;
+import org.springframework.integration.zmq.support.ZmqMessageConverter;
+import org.springframework.jmx.export.annotation.ManagedAttribute;
+import org.springframework.jmx.export.annotation.ManagedOperation;
+import org.springframework.jmx.export.annotation.ManagedResource;
+import org.springframework.messaging.MessagingException;
+import org.springframework.util.Assert;
+
+/**
+ * Abstract class for ZMQ Message-Driven Channel Adapters.
+ *
+ * @author Subhobrata Dey
+ * @since 5.1
+ *
+ */
+@ManagedResource
+@IntegrationManagedResource
+public abstract class AbstractZmqMessageDrivenChannelAdapter extends MessageProducerSupport {
+
+	private final String url;
+
+	private final String clientId;
+
+	private AtomicReference<Topic> topic;
```

Comment:
Doesn't looks like we need all this concurrency overhead since we support only one topic.

I understand that you modeled these channel adapters after MQTT, but should we really poll all the logic from there?